Source/WebKit/WebProcess/WebCoreSupport/WebNotificationClient.cpp

```diff
@@ -119,9 +125,27 @@ NotificationClient::Permission WebNotificationClient::checkPermission(ScriptExec
     if (!origin)
         return NotificationClient::Permission::Denied;
 
-    NotificationClient::Permission resultPermission;
-    callOnMainRunLoopAndWait([&resultPermission, origin = origin->data().toString().isolatedCopy()] {
-        resultPermission = WebProcess::singleton().supplement<WebNotificationManager>()->policyForOrigin(origin);
+    auto resultPermission = NotificationClient::Permission::Default;
+    auto topOrigin = context->topOrigin().data();
+    callOnMainRunLoopAndWait([this, &resultPermission, topOrigin = crossThreadCopy(context->topOrigin().data()), securityOrigin = crossThreadCopy(origin->data())] {
+        auto page = m_page->corePage();
```

Comment:
Can't m_page be null? It seems this change would break our support for Notification.permission in ServiceWorker (@cdumez and @youennf can confirm), because ServiceWorkerGlobalScope does not have PermissionController, and it has NotificationClient.

Reply:
Yes, I believe theoretically that `m_page` can be null.

It is a good point that we should be careful about not breaking Notification.permission support in service workers. I would have expected us to have tests for this though.

Reply:
Those tests should be under http/tests/workers/service. Looks like we do have some coverage.

Reply:
Actually, none of the http/tests/workers/service tests seems to call `Notification.permission` *from* the service worker, only from the main page. I suspect we don't have test coverage for `Notification.permission` in service workers and like @szewai I suspect it is broken by your patch since service workers don't have a WebPage.

Sadly, I don't think it will be easy to fix until we have support for the generic Permissions.query() in (service) workers :( Permissions.query() currently requires a WebPage to function and that will probably have to change to support service workers.

Reply:
Two things:
1. Do we think the implementation before this patch would work for serviceWorkers? It yes, then maybe we can use that implementation if `m_page` is `null` and if not, then we use the code path in this patch. I could add an extra check in the code for these 2 cases.
2. The only reason `m_page` is used is to access the `WebPermissionController` object. Is there a way to access the controller without getting it from the page? If yes, then we can remove the use of m_page, get the controller a different way and then this implementation could work for service workers as well correct?

Reply:
Currently the service workers don't have a WebPermissionController since it is constructed by the WebPage and service workers don't have an associated WebPage (only a dummy WebCore::Page).

Yes, I think the implementation before this patch should work for service workers, though I wish we had layout tests coverage to confirm. Using the old code path when m_page is null wouldn't really be useful, would it? We would still need to maintain the old code path, just for service workers (with very little or no test coverage of that code path). The whole point of this patch was to have a single code path and get rid of the WebProcess cache I think.

Reply:
We put this patch on hold since there was no support for service workers in Permissions::query. Since there is support now (after https://github.com/WebKit/WebKit/pull/3570), we can do the refactoring now.